File: android/src/com/dontflymoney/view/SettingsActivity.java
package com.dontflymoney.view;

import org.json.JSONException;
import org.json.JSONObject;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.DialogInterface.OnClickListener;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.CheckBox;

import com.dontflymoney.api.Request;
import com.dontflymoney.api.Step;

public class SettingsActivity extends SmartActivity
{
	CheckBox useCategories;
	
	public SettingsActivity()
	{
		init(this, R.layout.activity_settings, R.menu.settings);
	}
	
	@Override
	protected void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		populateScreen();
	}

	private void populateScreen()
	{
		getFields();
		getCurrentSettings();
	}
	
	private void getFields()
	{
		useCategories = (CheckBox) findViewById(R.id.use_categories);
	}

	private void getCurrentSettings()
	{
		Request request = new Request(this, "Users/GetConfig");
		request.AddParameter("ticket", Authentication.Get());
		request.Get(Step.Populate);
	}
	
	public void saveSettings(View view)
	{
		Request request = new Request(this, "Users/SaveConfig");
		request.AddParameter("ticket", Authentication.Get());
		request.AddParameter("UseCategories", useCategories.isChecked());
		request.Post(Step.Recording);
	}
	
	

	@Override
	protected void HandleSuccess(JSONObject data, Step step) throws JSONException
	{
		switch (step) {
			case Populate: {
				populateScreen(data);
				break;
			}
			case Recording: {
				back();
				break;
			}
			default: {
				alertError(R.string.this_is_not_happening);
				break;
			}
		}
		
	}
	
	
	private void populateScreen(JSONObject data) throws JSONException
	{
		boolean currentUseCategories = data.getBoolean("UseCategories");
		useCategories.setChecked(currentUseCategories);
	}
	
	
	private void back()
	{
		Bundle extras = getIntent().getExtras();
		final Class<?> parent = (Class<?>)extras.get("parent");
		
		new AlertDialog.Builder(this)
			.setTitle(R.string.title_activity_settings)
			.setMessage(R.string.settings_saved)
			.setPositiveButton(R.string.alert_button, new OnClickListener(){
				@Override
				public void onClick(DialogInterface dialog, int which) {
					redirect(parent);
				}
	    	})
			.show();
	}
	
	
	
	public void refresh(MenuItem menuItem) {
		populateScreen();
	}	
	
}

File: android/DFM/src/main/kotlin/com/darakeon/dfm/settings/SettingsActivity.kt
package com.darakeon.dfm.settings

import android.app.AlertDialog
import android.os.Bundle
import android.view.View
import com.darakeon.dfm.R
import com.darakeon.dfm.api.InternalRequest
import com.darakeon.dfm.auth.getAuth
import com.darakeon.dfm.base.SmartActivity
import com.darakeon.dfm.extensions.ON_CLICK
import com.darakeon.dfm.extensions.redirectWithExtras
import kotlinx.android.synthetic.main.settings.move_check
import kotlinx.android.synthetic.main.settings.use_categories
import org.json.JSONObject

class SettingsActivity : SmartActivity<SettingsStatic>(SettingsStatic) {

	override fun contentView(): Int = R.layout.settings


	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		if (rotated && succeeded) {
			use_categories.isChecked = static.useCategories
			move_check.isChecked = static.moveCheck
		} else {
			getCurrentSettings()
		}
	}

	private fun getCurrentSettings() {
		val request = InternalRequest(
			this, "Users/GetConfig", { d -> populateScreen(d) }
		)
		request.addParameter("ticket", getAuth())
		request.get()
	}

	fun saveSettings(@Suppress(ON_CLICK) view: View) {
		val request = InternalRequest(
			this, "Users/SaveConfig", { back() }
		)
		request.addParameter("ticket", getAuth())
		request.addParameter("UseCategories", use_categories.isChecked)
		request.addParameter("MoveCheck", move_check.isChecked)
		request.post()
	}


	private fun populateScreen(data: JSONObject) {
		static.useCategories = data.getBoolean("UseCategories")
		use_categories.isChecked = static.useCategories

		static.moveCheck = data.getBoolean("MoveCheck")
		move_check.isChecked = static.moveCheck
	}


	private fun back() {
		AlertDialog.Builder(this)
			.setTitle(R.string.title_activity_settings)
			.setMessage(R.string.settings_saved)
			.setPositiveButton(R.string.ok_button) { _, _ -> redirectWithExtras() }
			.show()
	}


}
